torchrl/modules/functional_modules.py

```diff
@@ -0,0 +1,227 @@
+                f"inputs, or you are trying to vmap over a function with no inputs. "
+                f"The latter is unsupported."
+            )
+
+        flat_args, args_spec = tree_flatten(args)
+        flat_in_dims = _broadcast_to_and_flatten(in_dims, args_spec)
+        if flat_in_dims is None:
+            raise ValueError(
+                f"vmap({_get_name(func)}, in_dims={in_dims}, ...)(<inputs>): "
+                f"in_dims is not compatible with the structure of `inputs`. "
+                f"in_dims has structure {tree_flatten(in_dims)[1]} but inputs "
+                f"has structure {args_spec}."
+            )
+
+        for i, (arg, in_dim) in enumerate(zip(flat_args, flat_in_dims)):
+            if not isinstance(in_dim, int) and in_dim is not None:
+                raise ValueError(
+                    f"vmap({_get_name(func)}, in_dims={in_dims}, ...)(<inputs>): "
+                    f"Got in_dim={in_dim} for an input but in_dim must be either "
+                    f"an integer dimension or None."
+                )
+            if isinstance(in_dim, int) and not isinstance(
+                arg, (Tensor, TensorDictBase)
+            ):
+                raise ValueError(
+                    f"vmap({_get_name(func)}, in_dims={in_dims}, ...)(<inputs>): "
+                    f"Got in_dim={in_dim} for an input but the input is of type "
+                    f"{type(arg)}. We cannot vmap over non-Tensor arguments, "
+                    f"please use None as the respective in_dim"
+                )
+            if in_dim is not None and (in_dim < -arg.dim() or in_dim >= arg.dim()):
+                raise ValueError(
+                    f"vmap({_get_name(func)}, in_dims={in_dims}, ...)(<inputs>): "
+                    f"Got in_dim={in_dim} for some input, but that input is a Tensor "
+                    f"of dimensionality {arg.dim()} so expected in_dim to satisfy "
+                    f"-{arg.dim()} <= in_dim < {arg.dim()}."
+                )
+            if in_dim is not None and in_dim < 0:
+                flat_in_dims[i] = in_dim % arg.dim()
+
+        return (
+            _validate_and_get_batch_size(flat_in_dims, flat_args),
+            flat_in_dims,
+            flat_args,
+            args_spec,
+        )
+
+    functorch._src.vmap._process_batched_inputs = _process_batched_inputs
+
+    def _create_batched_inputs(flat_in_dims, flat_args, vmap_level: int, args_spec):
+        # See NOTE [Ignored _remove_batch_dim, _add_batch_dim]
+        batched_inputs = [
+            arg
+            if in_dim is None
+            else arg.apply(lambda _arg: _add_batch_dim(_arg, in_dim, vmap_level))
+            if isinstance(arg, TensorDictBase)
+            else _add_batch_dim(arg, in_dim, vmap_level)
+            for in_dim, arg in zip(flat_in_dims, flat_args)
+        ]
+        return tree_unflatten(batched_inputs, args_spec)
+
+    functorch._src.vmap._create_batched_inputs = _create_batched_inputs
```

Comment:
The usual way to get a custom data structure to work with vmap is to register pytree definitions for it. This consists of writing two functions, a "flatten" (which, given a Data Structure, returns a list containing all the elements and some metadata), and an "unflatten" (which, given a list and some metadata reconstructs the Data Structure).

For example, [here is the definition for dict](https://github.com/pytorch/pytorch/blob/5e9c26c8e25e5d5be18ff98b7808b674b1e7a0a5/torch/utils/_pytree.py#L47-L51).

Any Data Structure that has a pytree definition automatically works with vmap. Does it make sense to do this for TensorDict?

Reply:
Sure I should try that out! Thanks for the tip